docs/docs/typelevel.md

```diff
@@ -0,0 +1,498 @@
+the use of quotes and rely on lifting to go from `Int` to
+`Expr[Int]`. So the following would also be correct and would return a
+constant without relying on constant folding:
+
+    inline def toInt[N <: Nat]: Int = ~{
+      case N =: Z => 0
+      case N =:= S[type N1] => 1 + toInt[N1]
+    }
+
+### Example: HLists
+
+Consider the standard definition of `HList`s:
+
+    class HList
+
+    case class HCons[X, Xs <: HList](hd: X, tl: Xs) extends HList
+    case object HNil extends HList
+
+A type-level length function on HLists can be written as follows:
+
+    type Length[Xs <: HList] = {
+      case Xs =:= HNil => Z
+      case Xs =:= HCons[type X, type Xs1] => S[Length[Xs1]]
+    }
+
+If we don't want to introduce the type variable `X`, which is unused, we could
+also formulate the type pattern with `<:<` instead of `=:=`:
+
+    type Length[Xs <: HList] = {
+      case Xs <:< HNil => Z
+      case Xs <:< HCons[_, type Xs1] => S[Length[Xs1]]
+    }
+
+Here's a `Concat` type with associated `concat` method:
+
+    type Concat[Xs <: HList, Ys <: HList] = {
+      case Xs =:= HNil => Ys
+      case Xs =:= HCons[type X, type Xs1] => HCons[X, Concat[Xs1, Ys]]
+    }
+
+    inline def concat[Xs <: HList, Ys <: HList](xs: Xs, ys: Ys): Concat[Xs, Ys] = ~ {
+      case Xs =:= HNil => 'ys
+      case Xs =:= HCons[type X, type Xs1] => '(xs.hd :: concat(xs.tl, ys))
+    }
+
+### Typechecking Query Conditionals
+
+Why does the last case of `concat` typecheck, given that `xs` is of
+type `Xs`, which does not have `hd` or `tl` fields? Roughly, it
+typechecks because `=:=` (and `<:<`) will inherit from
+`ImplicitConverter` so in the presence of an implicit search result of
+`T =:= U` or `T <:< U` we also get an implicit conversion from `T` to
+`U`. But we still need to explain why a search for `Xs =:= HCons[type X, type Xs1]`
+will succeed in the right hand side of the case. What happens in detail is this:
+
+When type checking a query case, the type pattern is type-checked
+according to the usual rules. Then the guard and right-hand side are type-checked
+under the assumption that an implicit value of the type pattern is available.
+For instance, when checking
+
+    case T => E
+
+`E` is typechecked in an environment that contains a definition like the following.
+
+    inline implicit def $ev: T = implicitly[T]
+
+Here, `$ev` stands for a compiler-defined fresh-name. The definition is declared `inline`
+so that it can be used in quoted as well as unquoted code.
+
+### Binding Implicit Condition Results
+
+The implicit definition backing a query pattern can be bound to a name `x` using the syntax
+
+    case x: T => ...
+
+Using this device we can make the conversions in the `concat` function above explicit:
+
+    inline def concat[Xs <: HList, Ys <: HList](xs: Xs, ys: Ys): Concat[Xs, Ys] = ~ {
+      case Xs =:= HNil => 'ys
+      case toCons: Xs =:= HCons[type X, type Xs1] => '(toCons(xs).hd :: concat(toCons(xs).tl, ys))
+    }
```

Comment:
It's not clear to me how the typer checks that the result type is `Concat[Xs, Ys]` at the definition site of `concat`?

Reply:
Very good question. I added a section explaining this. Essentially, we use only approximate typing for definitions and do the full typecheck at all expansion sites.